### ISSUE TITLE
Array.map typechecking issues, doesn't typecheck what the .map callback returns

### ISSUE THREAD

**Issue by artisologic** [microsoft/TypeScript#31425]
**TypeScript Version:**  reproducible in at least `3.3.3333` and above

**Search Terms:** Array map generic extra property

**Code**

```ts
type Token = {
    value: string
}

const things = ['one', 'two'];

// using .map generic — doesn't fail typechecking
function getTokens1() {
    return things.map<Token>(thing => ({
        value: thing,
        // should say `somethingElse` does not exist on type `Token`
        somethingElse: true
    }));
}

// using return type — doesn't fail typechecking
function getTokens2(): Token[] {
    const tokens = things.map(thing => ({
        value: thing,
        // should say `somethingElse` does not exist on type `Token`
        somethingElse: true
    }));

    // type of tokens is reported as `{ value: string; somethingElse: boolean; }[]`
    // so how could it be compatible with `Token[]`?
    return tokens;
}


// following use cases as working proofs and potential workarounds

// fails typechecking correctly
const token: Token = {
    value: 'one',
    somethingElse: true // `somethingElse` does not exist on type `Token`
};

// fails typechecking correctly — potential workaround
function getTokens3() {
    // note the `Token` inline return type
    return things.map((thing): Token => ({
        value: thing,
        somethingElse: true // `somethingElse` does not exist on type `Token`
    }));
}
```

**Expected behavior:**

I would expect in `getTokens1` using the `.map<U>` generic form that it would typecheck what `.map` returns. Similarly, I would expect in `getTokens2` that it would complain as I am precising the return type of the function and the inference itself seems to show an incompatible type.

**Actual behavior:**

It doesn't seem to typecheck properly when using `Array.map`.

**Playground Link:** [See above example in playground](https://www.typescriptlang.org/play/index.html#src=type%20Token%20%3D%20%7B%0D%0A%20%20%20%20value%3A%20string%0D%0A%7D%0D%0A%0D%0Aconst%20things%20%3D%20%5B'one'%2C%20'two'%5D%3B%0D%0A%0D%0A%2F%2F%20using%20.map%20generic%20%E2%80%94%20doesn't%20fail%20typechecking%0D%0Afunction%20getTokens1()%20%7B%0D%0A%20%20%20%20return%20things.map%3CToken%3E(thing%20%3D%3E%20(%7B%0D%0A%20%20%20%20%20%20%20%20value%3A%20thing%2C%0D%0A%20%20%20%20%20%20%20%20%2F%2F%20should%20say%20%60somethingElse%60%20does%20not%20exist%20on%20type%20%60Token%60%0D%0A%20%20%20%20%20%20%20%20somethingElse%3A%20true%0D%0A%20%20%20%20%7D))%3B%0D%0A%7D%0D%0A%0D%0A%2F%2F%20using%20return%20type%20%E2%80%94%20doesn't%20fail%20typechecking%0D%0Afunction%20getTokens2()%3A%20Token%5B%5D%20%7B%0D%0A%20%20%20%20const%20tokens%20%3D%20things.map(thing%20%3D%3E%20(%7B%0D%0A%20%20%20%20%20%20%20%20value%3A%20thing%2C%0D%0A%20%20%20%20%20%20%20%20%2F%2F%20should%20say%20%60somethingElse%60%20does%20not%20exist%20on%20type%20%60Token%60%0D%0A%20%20%20%20%20%20%20%20somethingElse%3A%20true%0D%0A%20%20%20%20%7D))%3B%0D%0A%0D%0A%20%20%20%20%2F%2F%20type%20of%20tokens%20is%20reported%20as%20%60%7B%20value%3A%20string%3B%20somethingElse%3A%20boolean%3B%20%7D%5B%5D%60%0D%0A%20%20%20%20%2F%2F%20so%20how%20could%20it%20be%20compatible%20with%20%60Token%5B%5D%60%3F%0D%0A%20%20%20%20return%20tokens%3B%0D%0A%7D%0D%0A%0D%0A%0D%0A%2F%2F%20following%20use%20cases%20as%20working%20proofs%20and%20potential%20workarounds%0D%0A%0D%0A%2F%2F%20fails%20typechecking%20correctly%0D%0Aconst%20token%3A%20Token%20%3D%20%7B%0D%0A%20%20%20%20value%3A%20'one'%2C%0D%0A%20%20%20%20somethingElse%3A%20true%20%2F%2F%20%60somethingElse%60%20does%20not%20exist%20on%20type%20%60Token%60%0D%0A%7D%3B%0D%0A%0D%0A%2F%2F%20fails%20typechecking%20correctly%20%E2%80%94%20potential%20workaround%0D%0Afunction%20getTokens3()%20%7B%0D%0A%20%20%20%20%2F%2F%20note%20the%20%60Token%60%20inline%20return%20type%0D%0A%20%20%20%20return%20things.map((thing)%3A%20Token%20%3D%3E%20(%7B%0D%0A%20%20%20%20%20%20%20%20value%3A%20thing%2C%0D%0A%20%20%20%20%20%20%20%20somethingElse%3A%20true%20%2F%2F%20%60somethingElse%60%20does%20not%20exist%20on%20type%20%60Token%60%0D%0A%20%20%20%20%7D))%3B%0D%0A%7D)

**Related Issues:**

**Comment by artisologic** [comment]
https://medium.com/@lemoine.benoit/why-does-typescript-sometimes-fails-to-type-check-extra-properties-fd230ebbc295

Not a bug then?

**Comment by fatcerberus** [comment]
I’m not sure why the first case isn’t an error (it seems like it *should* be, you’ve explicitly provided the result type for `map` and the object literal is still fresh when returned), but the second case is expected.  See my comment here re: structural typing: https://github.com/microsoft/TypeScript/issues/31412#issuecomment-493094951

**Comment by RyanCavanaugh** [comment]
Duplicate #241
